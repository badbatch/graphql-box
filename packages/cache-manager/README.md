# @graphql-box/cache-manager

The GraphQL Box cache manager module.

[![npm version](https://badge.fury.io/js/%40graphql-box%2Fcache-manager.svg)](https://badge.fury.io/js/%40graphql-box%2Fcache-manager)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)

## Installation

```bash
yarn add @graphql-box/cache-manager
```

## Usage

Read the GraphQL Box [usage guide](../../README.md#usage).

## Documentation

Read the module API [documentation](docs/README.md).
