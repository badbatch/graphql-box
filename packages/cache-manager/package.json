{
  "name": "@graphql-box/cache-manager",
  "description": "The GraphQL Box cache manager module.",
  "version": "5.4.14",
  "author": "miami-man",
  "license": "MIT",
  "homepage": "https://github.com/badbatch/graphql-box",
  "repository": {
    "directory": "packages/cache-manager",
    "type": "git",
    "url": "https://github.com/badbatch/graphql-box"
  },
  "bugs": "https://github.com/badbatch/graphql-box/issues",
  "type": "module",
  "main": "./dist/cjs/index.cjs",
  "module": "./dist/esm/index.mjs",
  "types": "./dist/types/cjs/index.d.cts",
  "imports": {
    "#*": "./src/*"
  },
  "exports": {
    ".": {
      "types": {
        "import": "./dist/types/esm/index.d.ts",
        "require": "./dist/types/cjs/index.d.cts"
      },
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.cjs"
    }
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "pnpm run clean:dist && pnpm run compile",
    "clean:deps": "del-cli ./node_modules",
    "clean:dist": "del-cli ./dist",
    "compile": "pnpm run /^compile:.*/",
    "compile:cjs": "MODULE_SYSTEM=cjs rollup -c ../../rollup.config.mjs",
    "compile:esm": "rollup -c ../../rollup.config.mjs",
    "compile:types": "tsc --project ./tsconfig.build.json && cts-types build dist/types/esm dist/types/cjs"
  },
  "dependencies": {
    "@graphql-box/core": "workspace:*",
    "@graphql-box/helpers": "workspace:*",
    "@types/lodash-es": "^4.17.12",
    "cacheability": "^4.0.30",
    "core-js": "^3.43.0",
    "lodash-es": "^4.17.21"
  },
  "peerDependencies": {
    "@cachemap/core": "<6",
    "graphql": "<17"
  },
  "devDependencies": {
    "@cachemap/core": "^5.2.8",
    "@cachemap/map": "^5.0.10",
    "@graphql-box/test-utils": "workspace:*",
    "@jest/globals": "^29.3.1",
    "cts-types": "^0.0.10",
    "del-cli": "^6.0.0",
    "graphql": "^16.11.0"
  },
  "keywords": [
    "client",
    "graphql",
    "graphql-box",
    "isomorphic",
    "server"
  ]
}
