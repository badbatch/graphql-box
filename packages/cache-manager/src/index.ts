export { CACHE_ENTRY_ADDED, CACHE_ENTRY_QUERIED, PARTIAL_QUERY_COMPILED } from "./consts";
export * from "./defs";
export { default, CacheManager } from "./main";
