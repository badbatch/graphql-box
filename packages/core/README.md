# @graphql-box/core

The GraphQL Box core module.

[![npm version](https://badge.fury.io/js/%40graphql-box%2Fcore.svg)](https://badge.fury.io/js/%40graphql-box%2Fcore)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)

## Installation

```bash
yarn add @graphql-box/core
```

## Usage

Read the GraphQL Box [usage guide](../../README.md#usage).

## Documentation

Read the module API [documentation](docs/README.md).
