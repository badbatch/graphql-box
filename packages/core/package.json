{
  "name": "@graphql-box/core",
  "version": "4.1.6",
  "description": "The GraphQL Box core module.",
  "author": "Dylan Aubrey <dylanaubrey@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/badbatch/graphql-box",
  "bugs": {
    "url": "https://github.com/badbatch/graphql-box/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/badbatch/graphql-box",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "main": "lib/main/index.js",
  "module": "lib/module/index.mjs",
  "browser": "lib/browser/index.js",
  "types": "lib/types/index.d.ts",
  "keywords": [
    "graphql-box",
    "graphql",
    "client",
    "server",
    "isomorphic"
  ],
  "scripts": {
    "compile": "concurrently npm:compile:*",
    "compile-watch": "concurrently npm:compile-watch:*",
    "compile-watch:main": "yarn run compile:main --watch",
    "compile-watch:types": "yarn run compile:types --watch",
    "compile:browser": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js",
    "compile:main": "cross-env BABEL_ENV=main babel ./src --out-dir ./lib/main --extensions \".ts\" --source-maps --config-file ../../babel.config.js",
    "compile:module": "cross-env BABEL_ENV=module babel ./src --out-dir ./lib/module --extensions \".ts\" --out-file-extension \".mjs\" --source-maps --config-file ../../babel.config.js",
    "compile:types": "tsc --build",
    "watch": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js -w"
  },
  "dependencies": {
    "serialize-error": "8.1.0"
  },
  "peerDependencies": {
    "@babel/runtime": "< 8",
    "core-js": "< 4",
    "lodash": "< 5"
  }
}