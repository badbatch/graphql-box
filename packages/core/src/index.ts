import * as coreDefs from "./defs";
import Core from "./main";

export {
  Core as default,
  coreDefs,
};
