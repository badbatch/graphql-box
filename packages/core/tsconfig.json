{
  "extends": "../tsconfig.settings.json",
  "compilerOptions": {
    "outDir": "lib/types",
    "rootDir": "src"
  },
  "include": [
    "src"
  ]
}