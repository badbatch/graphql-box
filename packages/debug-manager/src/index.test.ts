import Cachemap from "@cachemap/core";
import map from "@cachemap/map";
import { CACHE_ENTRY_ADDED, CACHE_ENTRY_QUERIED, CacheManager } from "@graphql-box/cache-manager";
import { DebugManagerDef, DEFAULT_TYPE_ID_KEY, PlainObjectMap } from "@graphql-box/core";
import {
  getRequestContext,
  getRequestData,
  parsedRequests,
  requestFieldTypeMaps,
  responses,
} from "@graphql-box/test-utils";
import { DebugManager } from ".";

const { performance } = window;

describe("@graphql-box/debug-manager >>", () => {
  const realDateNow = Date.now.bind(global.Date);
  let debugManager: DebugManagerDef;

  beforeAll(() => {
    global.Date.now = jest.fn().mockReturnValue(Date.parse("June 6, 1979 GMT"));
  });

  afterAll(() => {
    global.Date.now = realDateNow;
  });

  describe("CacheManager >> CACHE_ENTRY_ADDED >>", () => {
    const response: PlainObjectMap[] = [];

    beforeAll(async () => {
      debugManager = await DebugManager.init({ name: "CLIENT", performance });

      // @ts-ignore
      jest.spyOn(debugManager._performance, "now").mockReturnValue(0);

      debugManager.on(CACHE_ENTRY_ADDED, (payload: PlainObjectMap) => {
        response.push(payload);
      });

      const cacheManager = await CacheManager.init({
        cache: await Cachemap.init({
          name: "cachemap",
          store: map(),
        }),
        cascadeCacheControl: true,
        typeIDKey: DEFAULT_TYPE_ID_KEY,
      });

      const requestData = getRequestData(parsedRequests.singleTypeQuery);

      await cacheManager.resolveQuery(
        requestData,
        requestData,
        responses.singleTypeQuery,
        { awaitDataCaching: true },
        getRequestContext({ debugManager, fieldTypeMap: requestFieldTypeMaps.singleTypeQuery }),
      );
    });

    it("correct data emitted", () => {
      expect(response).toMatchSnapshot();
    });
  });

  describe("CacheManager >> CACHE_ENTRY_QUERIED >>", () => {
    const response: PlainObjectMap[] = [];

    beforeAll(async () => {
      debugManager = await DebugManager.init({ name: "CLIENT", performance });

      // @ts-ignore
      jest.spyOn(debugManager._performance, "now").mockReturnValue(0);

      debugManager.on(CACHE_ENTRY_QUERIED, (payload: PlainObjectMap) => {
        response.push(payload);
      });

      // @ts-ignore
      jest.spyOn(CacheManager, "_isValid").mockReturnValue(true);

      const cacheManager = await CacheManager.init({
        cache: await Cachemap.init({
          name: "cachemap",
          store: map(),
        }),
        cascadeCacheControl: true,
        typeIDKey: DEFAULT_TYPE_ID_KEY,
      });

      const requestData = getRequestData(parsedRequests.singleTypeQuerySet.initial);

      await cacheManager.resolveQuery(
        requestData,
        requestData,
        responses.singleTypeQuerySet.initial,
        { awaitDataCaching: true },
        getRequestContext({ fieldTypeMap: requestFieldTypeMaps.singleTypeQuery }),
      );

      await cacheManager.analyzeQuery(
        getRequestData(parsedRequests.singleTypeQuery),
        { awaitDataCaching: true },
        getRequestContext({ debugManager, fieldTypeMap: requestFieldTypeMaps.singleTypeQuery }),
      );
    });

    it("correct data emitted", () => {
      expect(response).toMatchSnapshot();
    });
  });
});
