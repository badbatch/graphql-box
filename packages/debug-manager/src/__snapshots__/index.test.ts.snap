// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@graphql-box/debug-manager >> CacheManager >> CACHE_ENTRY_ADDED >> correct data emitted 1`] = `
Array [
  Object {
    "cacheType": "requestFieldPaths",
    "cachemapOptions": Object {
      "cacheHeaders": Object {
        "cacheControl": "public, max-age=5",
      },
      "tag": undefined,
    },
    "context": Object {
      "boxID": "123456789",
      "fieldTypeMap": Map {
        "query.organization" => Object {
          "hasArguments": true,
          "hasDirectives": false,
          "isEntity": true,
          "isInterface": false,
          "isUnion": false,
          "possibleTypes": Array [],
          "typeIDValue": undefined,
          "typeName": "Organization",
        },
      },
      "operation": "query",
      "operationName": "",
      "queryFiltered": false,
      "request": "",
    },
    "debuggerName": "CLIENT",
    "hash": "5256db516a2438a7e5488a00413f4809",
    "options": Object {
      "awaitDataCaching": true,
    },
    "stats": Object {
      "duration": 0,
      "endTime": 0,
      "startTime": 0,
    },
    "value": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  Object {
    "cacheType": "dataEntities",
    "cachemapOptions": Object {
      "cacheHeaders": Object {
        "cacheControl": "public, max-age=5",
      },
      "tag": undefined,
    },
    "context": Object {
      "boxID": "123456789",
      "fieldTypeMap": Map {
        "query.organization" => Object {
          "hasArguments": true,
          "hasDirectives": false,
          "isEntity": true,
          "isInterface": false,
          "isUnion": false,
          "possibleTypes": Array [],
          "typeIDValue": undefined,
          "typeName": "Organization",
        },
      },
      "operation": "query",
      "operationName": "",
      "queryFiltered": false,
      "request": "",
    },
    "debuggerName": "CLIENT",
    "hash": "Organization::MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
    "options": Object {
      "awaitDataCaching": true,
    },
    "stats": Object {
      "duration": 0,
      "endTime": 0,
      "startTime": 0,
    },
    "value": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  Object {
    "cacheType": "queryResponses",
    "cachemapOptions": Object {
      "cacheHeaders": Object {
        "cacheControl": "public, max-age=5",
      },
      "tag": undefined,
    },
    "context": Object {
      "boxID": "123456789",
      "fieldTypeMap": Map {
        "query.organization" => Object {
          "hasArguments": true,
          "hasDirectives": false,
          "isEntity": true,
          "isInterface": false,
          "isUnion": false,
          "possibleTypes": Array [],
          "typeIDValue": undefined,
          "typeName": "Organization",
        },
      },
      "operation": "query",
      "operationName": "",
      "queryFiltered": false,
      "request": "",
    },
    "debuggerName": "CLIENT",
    "hash": "ea1cbc58eca798225b53f59d95ffb23a",
    "options": Object {
      "awaitDataCaching": true,
    },
    "stats": Object {
      "duration": 0,
      "endTime": 0,
      "startTime": 0,
    },
    "value": Object {
      "cacheMetadata": Object {
        "query": Object {
          "cacheControl": Object {
            "maxAge": 5,
            "public": true,
          },
          "etag": undefined,
          "ttl": 297475205000,
        },
        "query.organization": Object {
          "cacheControl": Object {
            "maxAge": 5,
            "public": true,
          },
          "etag": undefined,
          "ttl": 297475205000,
        },
      },
      "data": Object {
        "organization": Object {
          "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
          "email": "",
          "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
          "login": "facebook",
          "name": "Facebook",
          "url": "https://github.com/facebook",
        },
      },
    },
  },
]
`;

exports[`@graphql-box/debug-manager >> CacheManager >> CACHE_ENTRY_QUERIED >> correct data emitted 1`] = `
Array [
  Object {
    "cacheType": "requestFieldPaths",
    "context": Object {
      "boxID": "123456789",
      "fieldTypeMap": Map {
        "query.organization" => Object {
          "hasArguments": true,
          "hasDirectives": false,
          "isEntity": true,
          "isInterface": false,
          "isUnion": false,
          "possibleTypes": Array [],
          "typeIDValue": undefined,
          "typeName": "Organization",
        },
      },
      "operation": "query",
      "operationName": "",
      "queryFiltered": false,
      "request": "",
    },
    "debuggerName": "CLIENT",
    "hash": "5256db516a2438a7e5488a00413f4809",
    "options": Object {
      "awaitDataCaching": true,
    },
    "result": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
    },
    "stats": Object {
      "duration": 0,
      "endTime": 0,
      "startTime": 0,
    },
  },
  Object {
    "cacheType": "dataEntities",
    "context": Object {
      "boxID": "123456789",
      "fieldTypeMap": Map {
        "query.organization" => Object {
          "hasArguments": true,
          "hasDirectives": false,
          "isEntity": true,
          "isInterface": false,
          "isUnion": false,
          "possibleTypes": Array [],
          "typeIDValue": undefined,
          "typeName": "Organization",
        },
      },
      "operation": "query",
      "operationName": "",
      "queryFiltered": false,
      "request": "",
    },
    "debuggerName": "CLIENT",
    "hash": "Organization::MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
    "options": Object {
      "awaitDataCaching": true,
    },
    "result": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
    },
    "stats": Object {
      "duration": 0,
      "endTime": 0,
      "startTime": 0,
    },
  },
]
`;
